pkg/zsio/zeek/parser.go

```diff
@@ -22,60 +25,118 @@ type header struct {
 
 type parser struct {
 	header
-	unknown    int
-	needfields bool
-	needtypes  bool
-	descriptor *zson.Descriptor
-	addPath    bool
+	resolver    *resolver.Table
+	descriptors map[int]*zson.Descriptor
+	unknown     int  // Count of unknown directives
+
+	// Members below (plus header above) are all state related to
+	// parsing legacy zeek.  legacyVal indicates whether the last
+	// parsed directive was legacy or not.  As described in the
+	// zson spec, this governs whether values should be parsed as
+	// legacy or not.
+	// legacyDesc is a lazily-allocated Descriptor corresponding
+	// to the contents of the #fields and #types directives.
+	needfields  bool
+	needtypes   bool
+	legacyDesc *zson.Descriptor
+	legacyVal   bool
+	addPath     bool
 }
 
 var (
-	ErrBadRecordDef = errors.New("bad types/fields definition in zeek header")
-	ErrBadFormat    = errors.New("bad format")          //XXX
-	ErrBadValue     = errors.New("bad value")           //XXX
-	ErrBadEscape    = errors.New("bad escape sequence") //XXX
+	ErrBadRecordDef     = errors.New("bad types/fields definition in zeek header")
+	ErrBadFormat        = errors.New("bad format")          //XXX
+	ErrBadValue         = errors.New("bad value")           //XXX
+	ErrBadEscape        = errors.New("bad escape sequence") //XXX
+	ErrDescriptorExists = errors.New("descriptor already exists")
+	ErrInvalidDesc      = errors.New("invalid descriptor")
 )
 
-func newParser() *parser {
-	return &parser{header: header{separator: " "}}
+func newParser(r *resolver.Table) *parser {
+	return &parser{
+		header:      header{separator: " "},
+		resolver:    r,
+		descriptors: make(map[int]*zson.Descriptor),
+	}
 }
 
 func badfield(field string) error {
 	return fmt.Errorf("encountered bad header field %s parsing zeek log", field)
 }
 
-func (c *parser) parseFields(fields []string) error {
-	if len(c.columns) != len(fields) {
-		c.columns = make([]zeek.Column, len(fields))
-		c.needtypes = true
+func (p *parser) parseFields(fields []string) error {
+	if len(p.columns) != len(fields) {
+		p.columns = make([]zeek.Column, len(fields))
+		p.needtypes = true
```

Comment:
Nit: Mixing cosmetic changes like this rename in with the more substantive changes here makes more work for the reviewer. In the future, consider doing them in a separate PR.